Added Price to Catalog Items.

h-kataria opened the pull request:
Added Currency drop down and Price text field to Angular and non-angular Catalog Item editors. And update Catalog Item details screen t show the price information.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1602072

Angular Version Catalog Item editor
![angular_catalog_item_edit](https://user-images.githubusercontent.com/3450808/60275556-1bdf2200-98c8-11e9-97f7-b376ae8d4563.png)

Non- Angular version Catalog Item editor
![non-angular_catalog_item_edit](https://user-images.githubusercontent.com/3450808/60275578-24cff380-98c8-11e9-8347-fce7a0f91a7d.png)

Catalog Item summary screen
![catalog_item_with_pricing](https://user-images.githubusercontent.com/3450808/60275607-2bf70180-98c8-11e9-9156-59755e9fb8c1.png)

Price info on Catalog Service detail screen
![service_with_pricing](https://user-images.githubusercontent.com/3450808/60275617-31544c00-98c8-11e9-94e7-afc175841ae6.png)

Service that has no pricing info saved
![service_with_no_pricing](https://user-images.githubusercontent.com/3450808/60197391-47e99d00-980d-11e9-9e01-81471af234d5.png)

Catalog Item edit, show price as required field when Currency has been selected.
![price_required](https://user-images.githubusercontent.com/3450808/60275629-387b5a00-98c8-11e9-85f1-294022f2760a.png)

himdel commented on app/controllers/catalog_controller.rb:
```diff
@@ -1258,10 +1265,21 @@ def get_form_vars
+      @edit[:new][:currency] = params[:currency].to_i
+      @edit[:new][:code_currency] = code_currency_label(params[:currency])
+    end
+    @edit[:new][:price] = params[:price]
```
missing `if params[:price]`?

This is called both when saving from angular (`st_edit` with `?button=save` or `add`), and from non-angular `st_form_field_changed`.

himdel commented:
On the non-angular version, UAE Dirham is getting preselected (first option).
Maybe we should not preselect a currency? (Unless we have a user setting for a default currency somewhere?)

(Angular starts with "Nothing selected")

himdel commented on app/assets/javascripts/controllers/catalog/catalog_item_form_controller.js:
```diff
@@ -451,12 +456,14 @@ ManageIQ.angular.app.controller('catalogItemFormController', ['$scope', '$timeou
     $scope.$watch(fieldName, function(value) {
       vm.catalogItemModel[name + '_id'] = value ? value.id : '';
+      if (name ==  'currency')
+        vm.catalogItemModel.currency_name = _.find(vm.currencies, {id: value.id}).code;
```
This needs a check that `value` exists, I'm getting `Cannot read property 'id' of undefined` when loading the page (because no currency is selected).

himdel commented on app/views/catalog/_form_basic_info.html.haml:
```diff
@@ -62,6 +62,33 @@
+        %p.form-control-static
+          - currency = ChargebackRateDetailCurrency.currencies_for_select
+          = select_tag("currency",
+            options_for_select(currency, @edit[:new][:currency]),
```
aah, this is missing the usual `[["<Nothing>", nil]] + `

h-kataria commented:
@himdel please re-review

h-kataria commented:
@loicavenel cc

Loicavenel commented:
Ok, it is fine but I think we need to be more precise.. Price / Month

h-kataria commented:
@Loicavenel updated labels to `Price / Month ` , screenshots updated to reflect the change.

himdel commented:
Creating a Generic catalog item...

* there is no Price/Month field, only "Select Currency", the currency dropdown, and an unlabeled input on the next line (it's the price field, looks like you're only hiding the label)

* if I don't select a currency and try to save, I get an error:

```
 Error caught: [ActiveRecord::RecordNotFound] Couldn't find ChargebackRateDetailCurrency with 'id'=
/home/himdel/.rbenv/versions/2.4.4/lib/ruby/gems/2.4.0/gems/activerecord-5.1.7/lib/active_record/core.rb:189:in `find'
/home/himdel/manageiq-ui-classic/app/controllers/catalog_controller.rb:1284:in `code_currency_label'
/home/himdel/manageiq-ui-classic/app/controllers/catalog_controller.rb:1275:in `get_form_vars'
/home/himdel/manageiq-ui-classic/app/controllers/catalog_controller.rb:881:in `atomic_req_submit'
/home/himdel/manageiq-ui-classic/app/controllers/catalog_controller.rb:100:in `atomic_st_edit'
/home/himdel/manageiq-ui-classic/app/controllers/catalog_controller.rb:76:in `servicetemplate_edit'
```

As far as I understand, price & currency are not required fields, so, creating a catalog item without a price should still be working, right?

(If i do chose a currency, the price is required, that probably makes sense.)

himdel commented:
Creating an Ansible Playbook (angular) item works as expected, price is only required when a currency was chosen, saving seems to work fine.

The only issue I'm seeing is that `ggg` is a valid price now ;)

That is also present in the non-angular version.

h-kataria commented:
![Screenshot from 2019-07-01 19-29-58](https://user-images.githubusercontent.com/3450808/60472369-a8764100-9c36-11e9-8333-f47d85841fb9.png)

![Screenshot from 2019-07-01 19-29-15](https://user-images.githubusercontent.com/3450808/60472385-add38b80-9c36-11e9-9047-88439cfe4898.png)

@himdel addressed your feedback

himdel commented:
Trying to add a generic catalog item *without* price or currency now fails on `Price must be a numeric value`. Missing a currency present condition?

himdel commented:
* [x] can add a generic item without price or currency
* [x] can add a ansible playbook item without price or currency
* [x] can add a generic item with price and currency
* [x] can add a ansible playbook item with price and currency
* [x] validation fails for generic item with currency but no price
* [x] validation fails for generic item with currency but non-float price
* [x] validation fails for ansible playbook item with currency but no price
* [x] validation fails for ansible playbook item with currency but non-float price
* [x] can unset currency in generic
* [x] can unset currency in angular
* [x] can see currency & price in detail screen when set

Two potential issues when unsetting an already set currency:

* create a generic item with price and currency, save, edit, unset only currency.  
  The item will fail validation with `Price must be a numeric value`, which is surprising because there is a numeric price, just no currency.

* do the same in angular  
  Unsetting a currency works (does not clear the price, but does not show it in detail either),
  but gives an error in the console:

```
miq_debug.self-1351f771c2cd2fab0d83069fd3f62aadbb7effc11a01ed4942f52308793453be.js?body=1:30 TypeError: Cannot read property 'id' of null
    at catalog_item_form_controller.self-76ca03f2b26e2a8aa63b5e81271612a457ca56b30e7df8717628ad06eb469529.js?body=1:472
```

himdel commented on app/assets/javascripts/controllers/catalog/catalog_item_form_controller.js:
```diff
@@ -451,12 +462,16 @@ ManageIQ.angular.app.controller('catalogItemFormController', ['$scope', '$timeou
     var fieldName = 'vm._' + name;
     $scope.$watch(fieldName, function(value) {
       vm.catalogItemModel[name + '_id'] = value ? value.id : '';
+      if (name ==  'currency' && typeof value !== 'undefined') {
```
`if (name == 'currency' && ! value)` should fix the problem when unsetting currency. (It goes from undefined to a currency, but then it goes to null, not undefined, when unselecting.)

himdel commented:
LGTM,
removing currency autoremoves the price on save, 
removing just price warns about price required,
angular and ruby are consistent now :)
